UI/EncyclopediaDetailPanel.cpp

```diff
@@ -985,70 +1007,97 @@ void EncyclopediaDetailPanel::HandleSearchTextEntered() {
-                match_report += entry.second.first;
-                already_listed_results.insert({entry.second.second, entry.first});
+            if (GG::GUI::GetGUI()->ContainsWord(entry.second.first, word)) {
+                word_match_report.emplace(entry.second);
+                listed = true;
+                break;
             }
         }
-    }
-
-    // search for partial word matches: searched-for words that appear in the
-    // title text, not necessarily as a complete word
-    match_report += "\n" + UserString("ENC_SEARCH_PARTIAL_MATCHES") + "\n\n";
-    for (std::multimap<std::string, std::pair<std::string, std::string>>::value_type& entry : all_pedia_entries_list) {
-        std::set<std::pair<std::string, std::string>>::const_iterator dupe_it =
-            already_listed_results.find({entry.second.second, entry.first});
-        if (dupe_it != already_listed_results.end())
+        if (listed)
             continue;
 
+        // search for partial word matches: searched-for words that appear
+        // in the title text, not necessarily as a complete word
         for (const std::string& word : words_in_search_text) {
             // reject searches in text for words less than 3 characters
             if (word.size() < 3)
                 continue;
-            if (boost::icontains(entry.first, word)) {
-                match_report += entry.second.first;
-                already_listed_results.insert({entry.second.second, entry.first});
+            if (boost::icontains(entry.second.first, word)) {
+                partial_match_report.emplace(entry.second);
+                listed = true;
+                break;
             }
         }
-    }
-
-    // search for matches within article text
-    if (GetOptionsDB().Get<bool>("UI.encyclopedia.search.articles")) {
-        match_report += "\n" + UserString("ENC_SEARCH_ARTICLE_MATCHES") + "\n\n";
-        for (std::multimap<std::string, std::pair<std::string, std::string>>::value_type& entry : all_pedia_entries_list) {
-            std::set<std::pair<std::string, std::string>>::const_iterator dupe_it =
-                already_listed_results.find({entry.second.second, entry.first});
-            if (dupe_it != already_listed_results.end())
-                continue;
+        if (listed || !search_desc)
+            continue;
 
-            EncyclopediaArticle article_entry = GetPediaArticle(entry.second.second);
-            if (boost::icontains(UserString(article_entry.description), search_text)) {
-                match_report += entry.second.first;
-                already_listed_results.insert({entry.second.second, entry.first});
-                break;
-            }
+        // search for matches within article text
+        EncyclopediaArticle article_entry = GetPediaArticle(entry.first);
+        if (article_entry.description.empty()) {
+            // FIXME Articles generated through GetRefreshDetailPanelInfo
```

Comment:
I don't think searching dynamically generated article text is unnecessary. Most such articles are lists of stuff that would be found in another article anyway, or are lists of in-game objects that would be better listed directly in the search results themselves. Actually searching them all would also mean constructing them all, which would be very slow, and caching the results in some sort of search index, but that would be problematic because dynamically generated article contents can and do change.